spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerPersistentEntityImpl.java

```diff
@@ -94,12 +103,49 @@ private Expression detectExpression() {
 	public void addPersistentProperty(SpannerPersistentProperty property) {
 		addPersistentPropertyToPersistentEntity(property);
 		this.columnNames.add(property.getColumnName());
+
```

Comment:
Have you seen [`BasicPersistentEntity.verify(…)`](https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/mapping/model/BasicPersistentEntity.java#L422) as an overridable hook to validate the entity state? [Cassandra](https://github.com/spring-projects/spring-data-cassandra/blob/1d61f3c56d12fb707663d49fb470036ac12eddfa/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/mapping/BasicCassandraPersistentEntity.java#L161) uses this hook to perform various validations allowing to move the actual verification into a verifier object.

Reply:
Thank you! I did not know about this and will definitely take advantage.

Reply:
Perhaps I'm doing something wrong, but in some of my unit tests where the POJO has about 15 properties I'm seeing that even though I call getPersistentEntity (from AbstractMappingContext) once that several persistent entities are attempted to be created  sometimes on subsets of the properties, and each has their verify() called. Despite this , I do always get a full persistentEntity back at the end that has all the properties. The problem is that the verify() for those subsets-of-properties persistententities will fail. Does AbstractMappingContext do this to cache or optimize for performance?

Reply:
Not exactly sure what you mean. `MappingContext` adds typically an object graph of objects for a `PersistentEntity` that you request. Objects are derived from property types (i.e. having a `Person` class that has an `Address` property). Each entity is added through `addPersistentEntity(…)` and `.verify()` is only called after all properties were added.

Happy to look into your code if you can point me to a test that is able to reproduce the issue.

Reply:
Here is a test that has this behavior:

in class SpannerObjectMapperImpltTests, the test writeSomeColumnsTest. I then placed a breakpoint on BasicPersistentEntity's  verify() method (in this current version I have not overridden this method). When running the test in debug mode I hit that breakpoint in verify() multiple times and each time the BasicPersistentEntity has a different subset of TestEntity's properties (except the final time when all 16 appear).

Reply:
`verify()` is called on different entities (`Date`, `Timestamp`, `ByteArray`, `TestEntity`) only once.

One thing to note: `Date`, `Timestamp`, `ByteArray` look like data types that are natively supported by Spanner. Your mapping context should not represent these as persistent entities. We have an abstraction (`SimpleTypes`, see [`MongoSimpleTypes`](https://github.com/spring-projects/spring-data-mongodb/blob/148c4f9e249abc415c263f2e2d51ca5057763b48/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/mapping/MongoSimpleTypes.java) for implementation, [`MongoMappingContext`](https://github.com/spring-projects/spring-data-mongodb/blob/0be4d1345e80694bef0b8f6dc243570075b3ebd9/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/mapping/MongoMappingContext.java#L51) for usage) to control which types are considered simple ones. These types usually don't require a conversion but are understood natively by the driver.

Reply:
Ah thank you! I did not know that the AbstractMappingContext would be trying to create persistent entities for strange types automatically! 

Since we're not quite at the stage to support entities-within-entities, I'll override the `shouldCreatePersistentEntityFor` method and hardcode it to return false for the time being until we add that support.